### PR TITLE
[#IOPID-1866] delAssertionRef and delete installation repos and services

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

Porting logic for `LollipopRevokeQueue` and `Notifications` repo
Porting `deleteAssertionRefAssociation ` in `LollipopService`
Porting `delLollipopDataForUser` in `RedisSessionStorage`
Add notifications types

#### Motivation and Context

We are porting in this PR some functionalities needed to implement `acs` and `logout`

#### How Has This Been Tested?

unit test

#### Types of changes

- [X] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    };
+    mockSendMessage.mockResolvedValueOnce(expectedResponse);
+    const result = await pipe(
+      revokePreviousAssertionRef(anAssertionRef)({
+        lollipopRevokeQueueClient: mockQueueClient,
+      }),
+      TE.map((value) => expect(value).toEqual(expectedResponse)),
+    )();
+    expect(mockSendMessage).toBeCalledWith(expectedMessage);
+    expect(E.isRight(result)).toBeTruthy();
```
Comment (gquadrati) [inline apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts:35]: nitpick: we could define a `mockedDependencies` object, and we could check the result directly

```suggestion
    const mockedDependencies = {
      lollipopRevokeQueueClient: mockQueueClient,
    };
    const result = await pipe(
      mockedDependencies,
      revokePreviousAssertionRef(anAssertionRef),
    )();
    expect(mockSendMessage).toBeCalledWith(expectedMessage);
    expect(result).toEqual(E.right(expectedResponse));
```

File: apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  test("should returns an error if the queue client throws", async () => {
+    const expectedError = new Error("Network Error");
+    mockSendMessage.mockRejectedValue(expectedError);
+    const result = await pipe(
+      revokePreviousAssertionRef(anAssertionRef)({
+        lollipopRevokeQueueClient: mockQueueClient,
+      }),
+      TE.mapLeft((value) => expect(value).toEqual(expectedError)),
```
Comment (gquadrati) [inline apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts:45]: ditto

File: apps/session-manager/src/__mocks__/repositories/queue-client.mocks.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { QueueClient } from "@azure/storage-queue";
+import { vi } from "vitest";
+
+export const mockSendMessage = vi.fn();
```
Comment (gquadrati) [inline apps/session-manager/src/__mocks__/repositories/queue-client.mocks.ts:4]: It should be better to define Parameter and Response types, like [here](https://github.com/pagopa/io-auth-n-identity-domain/blob/465db3717f1a794ebdc09664bdc84d7707f5f489/apps/session-manager/src/__mocks__/services/lollipopService.mocks.ts#L12-L13)

File: apps/session-manager/src/services/lollipop.ts [hunk 4]
```diff
@@ -83,3 +86,58 @@ export const generateLCParams: (
...
+  Error,
+  boolean
+> = (fiscalCode, assertionRefToRevoke, eventName, eventMessage) => (deps) => {
+  // Sending a revoke message for assertionRef
+  // This operation is fire and forget
+  pipe(
+    LollipopRevokeRepo.revokePreviousAssertionRef(assertionRefToRevoke)(deps),
+    TE.mapLeft((err) => {
+      deps.appInsightsTelemetryClient?.trackEvent({
+        name: eventName,
```
Comment (gquadrati) [inline apps/session-manager/src/services/lollipop.ts:110]:
```suggestion
        name: eventName + ".revoke-assertion-ref",
```

File: apps/session-manager/src/services/lollipop.ts [hunk 5]
```diff
@@ -83,3 +86,58 @@ export const generateLCParams: (
...
+  )().catch(() => void 0); // This promise should never throw
+
+  return pipe(
+    RedisSessionStorageService.delLollipopDataForUser({
+      ...deps,
+      fiscalCode,
+    }),
+    TE.mapLeft((err) => {
+      deps.appInsightsTelemetryClient?.trackEvent({
+        name: eventName + ".delete",
```
Comment (gquadrati) [inline apps/session-manager/src/services/lollipop.ts:133]:
```suggestion
        name: eventName + ".delete-lollipop-data",
```

File: apps/session-manager/src/repositories/index.ts [hunk 6]
```diff
@@ -1,6 +1,15 @@
 import * as FnAppRepo from "./fn-app-api";
 import * as RedisRepo from "./redis";
-import * as LollipopApi from "./lollipop-api";
+import * as FnFastLoginRepo from "./lollipop-api";
```
Comment (gquadrati) [inline apps/session-manager/src/repositories/index.ts:3]: I don't think this rename is correct

File: apps/session-manager/src/services/__tests__/lollipop.spec.ts [hunk 7]
```diff
@@ -20,10 +22,16 @@ import {
   toNotFoundError,
   unauthorizedError,
 } from "../../models/domain-errors";
+import { RedisSessionStorageService } from "..";
+import { LollipopRevokeRepo } from "../../repositories";
+import { aFiscalCode } from "../../__mocks__/user.mocks";
+import { RedisClientSelectorType } from "../../types/redis";
+import { beforeEach } from "node:test";
```
Comment (gquadrati) [inline apps/session-manager/src/services/__tests__/lollipop.spec.ts:29]: Import it from "vitest"  like others

File: apps/session-manager/src/services/__tests__/lollipop.spec.ts [hunk 8]
```diff
@@ -105,3 +113,97 @@ describe("generateLCParams", () => {
...
+    )();
+
+    expect(mockRevokePreviousAssertionRef).toBeCalledWith(anAssertionRef);
+    expect(mockDelLollipopDataForUser).toBeCalledWith(
+      expect.objectContaining({ fiscalCode: aFiscalCode }),
+    );
+    expect(E.isLeft(result)).toBeTruthy();
+  });
+
+  test("should success if fire and forget revokePreviousAssertionRef fails", async () => {
```
Comment (gquadrati) [inline apps/session-manager/src/services/__tests__/lollipop.spec.ts:183]: Can we check if `trackEvent` is called?

File: apps/session-manager/src/types/notifications.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  "DeleteInstallation" = "DeleteInstallation",
+}
+
+export type NotificationMessageKind = t.TypeOf<typeof NotificationMessageKind>;
+export const NotificationMessageKind = enumType<NotificationMessageKindEnum>(
+  NotificationMessageKindEnum,
+  "NotificationMessageKind",
+);
+
+export enum KindEnum {
```
Comment (gquadrati) [inline apps/session-manager/src/types/notifications.ts:19]: I'd rather call it `DeleteInstallationKindEnum`

File: apps/session-manager/src/types/notifications.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  "DeleteInstallation" = "DeleteInstallation",
+}
+
+/**
+ * Message sent to the queue for a new Delete Installation event
+ */
+const DeleteInstallationMessage = t.type({
```
Comment (gquadrati) [inline apps/session-manager/src/types/notifications.ts:26]: The generated type is a `t.exact` of it

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: apps/session-manager/src/types/notifications.ts (hunk 10 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/types/notifications.ts]: Fixed in 4ede92b

File: apps/session-manager/src/types/notifications.ts (hunk 9 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/types/notifications.ts]: Fixed in 4ede92b

File: apps/session-manager/src/services/__tests__/lollipop.spec.ts (hunk 7 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/services/__tests__/lollipop.spec.ts]: Fixed in 4ede92b

File: apps/session-manager/src/repositories/index.ts (hunk 6 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/repositories/index.ts]: You are right, fixed in 4ede92b

File: apps/session-manager/src/services/lollipop.ts (hunk 4 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/services/lollipop.ts:110]: Why this change, will create a new event. Now in production there are 2 different points where the event is sent as `lollipop.error.acs`

File: apps/session-manager/src/__mocks__/repositories/queue-client.mocks.ts (hunk 3 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/__mocks__/repositories/queue-client.mocks.ts]: Great suggestion, fixed in e449053

File: apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts (hunk 1 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts]: Fixed in 21e33f0

File: apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts (hunk 2 above)
Comment (BurnedMarshal) [reply apps/session-manager/src/repositories/__tests__/lollipop-revoke-queue.spec.ts]: Fixed in 21e33f0

**Review by gquadrati — APPROVED** [review]
LGTM

**Pull request merged**